File: Chapter 10/C#/Section_10_3_Resilience_Decorators/Resilient/Infrastructure/ZebraPaymentAdapter.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;
using Polly;
using Polly.Retry;
using Chapter10.Resilience.Core.Ports;

namespace Chapter10.Resilience.Infrastructure.Adapters;

/// <summary>
/// THE INFRASTRUCTURE ADAPTER (The Implementation):
/// 
/// ARCHITECTURAL CRITIQUE:
/// This class encapsulates the Physical Resource Policy for the Zebra vendor.
/// By moving our SLA (Service Level Agreement) into named constants, we 
/// transform hidden magic numbers into a documented, tunable boundary.
/// The Core Application remains pure because the Polly retry policy is 
/// physically locked inside this adapter.
/// </summary>
public class ZebraPaymentAdapter : IPaymentGateway
{
    // --- THE PHYSICAL POLICY CONSTANTS (The SLA) ---
    // Connect/Request timeout: The absolute "Escape Hatch" for the thread.
    private const int TotalRequestTimeoutSec = 10;
    
    // Retry Policy constants
    private const int MaxRetryAttempts = 5;
    private const int InitialDelaySec = 2;
    private const int BackoffPower = 2;

    private readonly HttpClient _httpClient;
    private readonly AsyncRetryPolicy<bool> _retryPolicy;

    public ZebraPaymentAdapter(string baseUrl)
    {
        // Physical Bulkheading via HttpClient timeout configuration.
        _httpClient = new HttpClient 
        { 
            BaseAddress = new Uri(baseUrl),
            Timeout = TimeSpan.FromSeconds(TotalRequestTimeoutSec)
        };

        // THE SHIELD (Declarative Policy via Polly)
        // #SENIOR NOTE: Notice how the wait time scales exponentially: 2, 4, 8, 16, 32s.
        _retryPolicy = Policy<bool>
            .Handle<HttpRequestException>()
            .Or<TaskCanceledException>() // Explicitly catches the timeout from #A
            .WaitAndRetryAsync(MaxRetryAttempts, retryAttempt => 
                TimeSpan.FromSeconds(Math.Pow(BackoffPower, retryAttempt))
            );
    }

    public async Task<bool> ChargeAsync(decimal amount, string orderId, string idempotencyKey)
    {
        return await _retryPolicy.ExecuteAsync(async () =>
        {
            Console.WriteLine($"      [Zebra Adapter] Attempting Zebra Charge for {orderId}...");

            // HEADERS (Infrastructure Concern)
            _httpClient.DefaultRequestHeaders.Clear();
            _httpClient.DefaultRequestHeaders.Add("Idempotency-Key", idempotencyKey);

            var response = await _httpClient.PostAsJsonAsync("/charge", new {
                amount = amount,
                order_id = orderId
            });

            // This triggers the Handle<HttpRequestException> in the policy
            response.EnsureSuccessStatusCode();
            return true;
        });
    }
}

File: Chapter 10/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter10/section_10_3_resilience_decorators/resilient/infrastructure/adapters/ZebraPaymentAdapter.java
package com.grokking.resilience.infrastructure.adapters;

import com.grokking.resilience.core.ports.PaymentGateway;
import io.github.resilience4j.retry.Retry;
import io.github.resilience4j.retry.RetryConfig;
import java.time.Duration;
import java.util.function.Supplier;

/**
 * THE INFRASTRUCTURE ADAPTER:
 * * ARCHITECTURAL CRITIQUE:
 * This file encapsulates the Physical Resource Policy. By moving every 
 * configuration value into named constants, we transform a "hidden script" 
 * into a documented Service Level Agreement (SLA).
 */
public class ZebraPaymentAdapter implements PaymentGateway {

    // --- THE PHYSICAL POLICY CONSTANTS (The SLA) ---
    private static final int CONNECT_TIMEOUT_SECONDS = 2;
    private static final int READ_TIMEOUT_SECONDS = 8;
    
    private static final int MAX_RETRIES = 5;
    private static final int INITIAL_RETRY_DELAY_SECONDS = 2;
    private static final double BACKOFF_MULTIPLIER = 2.0;
    private static final int MAX_RETRY_DELAY_SECONDS = 10;

    private final String baseUrl;
    private final Retry retryShield;

    public ZebraPaymentAdapter(String baseUrl) {
        this.baseUrl = baseUrl;

        // THE SHIELD: Declarative Policy via Resilience4j
        // #SENIOR NOTE: Notice how we map the raw constants into 'Duration' objects 
        // here. This makes the configuration readable and type-safe.
        RetryConfig config = RetryConfig.custom()
                .maxAttempts(MAX_RETRIES)
                .waitDuration(Duration.ofSeconds(INITIAL_RETRY_DELAY_SECONDS))
                .backoffConfig(
                    io.github.resilience4j.core.IntervalFunction
                        .ofExponentialBackoff(
                            Duration.ofSeconds(INITIAL_RETRY_DELAY_SECONDS), 
                            BACKOFF_MULTIPLIER, 
                            Duration.ofSeconds(MAX_RETRY_DELAY_SECONDS)
                        )
                )
                .retryExceptions(Exception.class)
                .build();

        this.retryShield = Retry.of("zebra-charge-retry", config);
    }

    @Override
    public boolean charge(double amount, String orderId, String idempotencyKey) {
        Supplier<Boolean> resilientCall = Retry.decorateSupplier(retryShield, () -> {
            System.out.println("      [Zebra Adapter] Attempting Order " + orderId + "...");
            
            /* * TEACHING NOTE:
             * A real HttpClient would use CONNECT_TIMEOUT_SECONDS and 
             * READ_TIMEOUT_SECONDS to ensure we never hang the thread pool.
             */
            return true; 
        });

        return resilientCall.get();
    }
}